### PR TITLE
Run tests on ubuntu, windows and macos

### PR DESCRIPTION
**Pull request by kiljacken**
Make GitHub actions run our tests on all their currently supported platforms.

This will probably require some trial and error as this can't really be tested locally :)

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Just a little remainder (from my painful failure experiences):

bors look for status check of the action full name. So if we use `strategy` and `matrix` in our github action script, the `bors.toml` have to be changed to :

```toml
status = [
    "TypeScript",
    'Rust (macOS-latest)',
    'Rust (windows-2019)',
    'Rust (ubuntu-latest)'
]
```

**Comment by kiljacken** [comment]
Okay, that could've been way worse. Assuming bors agrees with my changes to it's config file this should be good to go.

This does make CI a bit slower, as compiler performance seems to be a bit worse on the windows workers. But all the workers run in parallel, so it's "only" 3-4 minutes slower overall, because the windows worker is straggling behind.

The macOS worker is surprisingly fast, around 2 minutes faster than the ubuntu build!

**Comment by edwin0cheng** [comment]
bors try

File: bors.toml [hunk 1]
```diff
@@ -1,2 +1,7 @@
-status = ["Rust", "TypeScript"]
+status = [
+    'Rust (ubuntu-latest)'
```
Comment (edwin0cheng) [inline bors.toml]: It is missing a comma, right ?

Comment (kiljacken) [reply bors.toml]: Yes, good eyes!

File: bors.toml [hunk 2]
```diff
@@ -1,2 +1,7 @@
...
+    'Rust (windows-latest)',
+    'Rust (macos-latest)',
+    "TypeScript"
+]
```
Comment (edwin0cheng) [inline bors.toml]: Maybe using single quote to make it consist to others ?

**Comment by lnicola** [comment]
IIRC MacOS has the toolchain already installed, but it's faster even disregarding that.

> 3-4 minutes slower overall

Maybe we could disable the "heavy" tests on Windows.

**Comment by flodiebold** [comment]
I think the idea at some point was that the platform-dependent parts are in different repositories, so this repository wouldn't need to be tested on every platform. Maybe we could just run the cargo-watch tests (and maybe the heavy tests, since they also tend to be more platform-dependent) on every platform? Don't know if that would help much :thinking:

**Comment by kiljacken** [comment]
Rebased on top of #2740.

Most of the time on windows come from installing the toolchain and compiling dependencies. The only configuration that currently might be slightly faster is only running cargo watch tests on windows. For the heavy tests we need to compile everything anyways so it'll perform basically the same as now.

**Comment by bjorn3** [comment]
Why is the `rust-docs` component being installed while the rustup profile is set to `minimal`? `rust-docs` is very slow to install on Windows due to the huge amount of files.

**Comment by kiljacken** [comment]
It seems that windows-latest comes pre-installed with rustup and older version of rust, so the components are already setup. Then when we update to latest stable it updates docs as well

**Comment by kiljacken** [comment]
Okay, removing `rust-docs` before updating to the latest stable is 20-30 seconds faster, but removing the component takes most of the total time saved. Does anybody know if there's a way to only update certain components?

**Comment by lnicola** [comment]
So now it spends 1.2 minutes removing the old `rust-docs`..

**Comment by bjorn3** [comment]
What about moving the `%HOME%\.rustup\toolchain\stable-x86_64-pc-windows-msvc` directory somewhere else instead, so that nothing gets deleted and rustup doesn't know that `rust-docs` was ever installed.

**Comment by kiljacken** [comment]
@bjorn3 That sounds about hacky enough to work, giving it a try

**Comment by kiljacken** [comment]
That seems to do the trick, now to see how it measures up.

**Comment by kiljacken** [comment]
That did improve the situation a bit Windows is now "only" two minutes slower than Linux, mostly down to the compilation speed difference (linux ~5m13s, windows ~7m36s)

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -20,6 +23,10 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v1
 
+      - name: Remove existing rust toolchain (Windows)
+        if: matrix.os == 'windows-latest'
+        run: Rename-Item C:\Users\runneradmin\.rustup\toolchains\stable-x86_64-pc-windows-msvc C:\Users\runneradmin\.rustup\toolchains\stable-x86_64-pc-windows-msvc.old
```
Comment (bjorn3) [inline .github/workflows/ci.yaml]: Maybe add a comment that is prevents installing `rust-docs`, which is slow, while not having to delete it, which is slow too?

Comment (kiljacken) [reply .github/workflows/ci.yaml]: Good idea.

**Comment by gilescope** [comment]
Nice work!
Are you planning on adding an upload task on this PR or is that better done as a separate PR?

**Comment by kiljacken** [comment]
@gilescope That's probably best left for another PR

File: bors.toml (hunk 1 above)
Comment (Arzte) [reply bors.toml]: This seems to still be missing a comma on the latest commit (e997749a9c2537b8bd4c894aa14a8db9158a1c73)

Comment (kiljacken) [reply bors.toml]: Must've missed a commit when bringing over changes from master, should be good now.

**Comment by edwin0cheng** [comment]
Let try it in bor-s.

bors try

**Comment by edwin0cheng** [comment]
bors try-

**Comment by edwin0cheng** [comment]
bors try

**Comment by edwin0cheng** [comment]
Seem like we can't run bors here now, because :

"OAuth tokens and GitHub Apps are restricted from editing the main.workflow file. This means it isn't possible to merge a Pull Request via an Action if it updates the main.workflow file I'm afraid."


https://github.community/t5/GitHub-Actions/GitHub-action-Resource-not-accessible-by-integration/m-p/19950/highlight/true#M75

**Comment by kiljacken** [comment]
@matklad Since bors is acting up, do you want to take it from here?

**Comment by gilescope** [comment]
@matklad thoughts on how we land this? Keen on not having to build rust-analyzer locally - then we have no need to install node.

**Comment by matklad** [comment]
We probably should move cache-cleaning to xtask thoguh, to not duplicate it across platforms

**Pull request merged**